### ISSUE TITLE
Total reformatting

### ISSUE THREAD

**Issue by snakers4** [snakers4/open_stt#17]
We have run some benchmarks with the `soundfile` library.
So far they seem very promising.

## OGG format benches

Compared time for several options on 10,000 samples:

- `wav` (from nvme) with `scipy.io.wavfile.read` - *~3 sec*;
- `ogg` (from nvme) with `soundfile` - *57 sec* - **best option**;
- `wav` (from hdd raid) with `scipy.io.wavfile.read`- *~3 min*;
- `ogg` (from nvme) to `wav` using `sox` , then read `wav` with `scipy` - *~20 min*;
- `wav`  (from hdd raid) with `librosa` *>30 min* (librosa normalizes);

```
import soundfile as sf
sound_array, sampling_rate = sf.read('test.ogg', dtype='int16')
```

## **We are considering for 1.2 (not soon anyway)**

- dropping `wav` format altogether
- dropping `mp3` format altogether
- converting everything to `ogg` and using `soundfile`
- maybe ditching the direct links (?)

## **Libraries I used / know**

- `scipy.io.wavfile.read` - the best option, but `wavs` take a lot of space. Cannot scale 10x more
- `mp3` decoding - also good, but would occupy additional CPU. All libraries are either slow or bulky (requires fiddling with temporary files, i.e. `soxi` - you have to be careful - you can exhaust your nvme drive resource)
- `torchaudio` - seems the most bulky library, also it used `ffmpeg` which is slower that `soxi`
- `librosa` - overall the slowest library

## **Motivation**

- Ogg takes 10x less space
- Less IO heavy load (but high IOPS nevertheless)
- Looks like `pysoundfile` does all this in memory

```
PySoundFile can read and write sound files. File reading/writing is supported through libsndfile, which is a free, cross-platform, open-source (LGPL) library for reading and writing many different sampled sound file formats that runs on many platforms including Windows, OS X, and Unix. It is accessed through CFFI, which is a foreign function interface for Python calling C code. CFFI is supported for CPython 2.6+, 3.x and PyPy 2.0+. PySoundFile represents audio data as NumPy arrays.
```

@buriy 
Maybe will be a good idea to do this when we will be doing a major annotation clean-up?

**Comment by snakers4** [comment]
So the question to the community is - do you see any drawbacks with such a migration?

**Comment by Islanna** [comment]
Also, we have an idea to read `mp3` from nvme, but `libsndfile` and `soundfile` lib don't support this format. 
According to docs of `libsndfile`: 
> I have decided that I will not be adding support for MPEG Layer 3 (commonly known as MP3) due to the patent issues surrounding this file format.

**Comment by mikhail-barg** [comment]
> converting everything to ogg and using soundfile

It would be nice to specify the codec you are planning to use (`ogg` is just a container). I hope it would be `opus`, correct?

From our old experiments we know that `mp3` causes 1-3% WER increase even in best quality, while converting files with `opus` does not cause this degradation and even provides for 1-2% WER decrease.

**Comment by snakers4** [comment]
> be opus, correct?

it is the default codec of choice but I am not sure which codec is used by `libsndfile` and `soundfile` and `soxi` by default (@Islanna you used `soxi`  to encode?)

**Comment by buriy** [comment]
@mikhail-barg thanks a lot for this insight!
Have you checked opus and mp3 reading speed? An old benchmark I found shows that opus is 2x-3x slower than mp3 to decompress, and so it might not work well for training. Then it only would suit well for storing data in an intermediate compressed format.

**Comment by mikhail-barg** [comment]
No, we haven't checked reading speed. Anyway it would probably depend on codec implementation and used interop library, so there would be no definite answer here.

**Comment by snakers4** [comment]
With enough CPU cores and an nvme(s) drive I believe at some point the advantages of compressed speech will overwhelm the drawbacks

Also I believe that will be definitely sticking to `.ogg` in future both for training
But that will happen during next large dataset growth phase

**Comment by snakers4** [comment]
@mikhail-barg 
Dug a bit deeper in the `ogg` encoding stuff

(i)
`pysoundfile` that uses [libsoundfile ](http://www.mega-nerd.com/libsndfile/) as a backend, but this backend supports only `ogg/vorbis`, not `ogg/opus` which is a separate codec

`pysoundfile`  seems to be the only python library on the market (correct me if I am wrong) that reads `.ogg` files in-memory, which is a big deal for large scale model training

So the question is, did you compare `ogg/opus` vs `ogg/vorbis`?

(ii)
There is another way - you could potentially use torchaudio CFFI [wrappers ](https://pytorch.org/audio/sox_effects.html) around sox, but I belive this is not a very good way

This [piece](https://www.quora.com/How-do-the-audio-formats-OPUS-and-OGG-compare-to-each-other) provides some clarity on the topic

(iii)
Also, how much time do your quality comparsion experiments take? Have you tried LPCnet [codec](https://people.xiph.org/~jm/demo/lpcnet_codec/) instead of `ogg/opus`?
Maybe for speech it is worth converting to lpc_net codec?

**Comment by snakers4** [comment]
@mikhail-barg 
(iv)
Also is this WER dependence on format somehow related to bitrate?
Which bitrates did you use?

**Comment by snakers4** [comment]
@mikhail-barg 
(v)
Also I wonder whether you tried `speex`
And which params do you use for opus, i.e. sample_rate / bitrate?

**Comment by mikhail-barg** [comment]
@snakers4 Sorry for the late answer, just noticed that your messages.

We've just used `opus` (via it's original `opusenc`/`opusdec` tools). We haven't tried `speex` and `vorbis` for these are officially inferior to `opus` at every possible combination, see [wikipedia](https://en.wikipedia.org/wiki/Opus_(audio_format)) and [this](http://opus-codec.org/comparison/).

I cannot say anything on the python support for opus, for we are not using python for speech corpuses processing. But as far as I can see, it should be possible to either use CLI `opusenc`/`opusedc` tools, or make native bindings to the underlying native library. [This](https://github.com/orion-labs/opuslib) seem to be an attempt to make latter.

**Comment by mikhail-barg** [comment]
> Also, how much time do your quality comparsion experiments take? 

We have done those experiments long ago, got the results and have been sticking to them since then. In general to make an experiment you have to have original uncompressed wavs for your corpus (that were never compressed with a lossy codec), compress those wavs with the codec of you choice, decompress back, and train the model from scratch on those decompressed wavs. So you may estimate the time required.

> Have you tried LPCnet codec instead of ogg/opus?

No we haven't. At the time we were conducting the experiment, there were no LPCnet codec )

> Maybe for speech it is worth converting to lpc_net codec?

It could be. See for example Results section [here](https://people.xiph.org/~jm/demo/lpcnet_codec/) or [this paper](https://arxiv.org/abs/1905.04628).

Still both of these are considering really low bitrate compression, so it's not clear if the results would stand at some less extreme bitrates.

> And which params do you use for opus, i.e. sample_rate / bitrate?

we are using `--hard-cbr --comp 10 --bitrate 48 --no-coupling --quiet`
note that `--no-coupling` is our custom patch for `opusenc` which prevents joint channels encoding. But given that your wavs are all single-channel, this should not affect your case.

**Comment by Worldexe** [comment]
> we are using --hard-cbr --comp 10 --bitrate 48 --no-coupling --quiet

Note that bitrate 48 is for two-channel data; single channel wavs are encoded with bitrate 24.

Also, that `--no-coupling` flag may already be included in `opusenc` somehow, I just not sure what the option name is.

**Comment by snakers4** [comment]
> Note that bitrate 48 is for two-channel data; single channel wavs are encoded with bitrate 

@Islanna 
is there any way to know which settings does libsndfile use?

**Comment by Islanna** [comment]
> is there any way to know which settings does libsndfile use?

Unfortunately, `libsndfile` docs lack info about the default settings for opus encoding. That's what I've managed to find out from the source code:
- Encoding/decoding via the original `libopus`; As far as I know `opusenc/dec` tools also depend on it;
- **Variable bit rate** encoding instead of constant bit rate(`--hard-cbr` option);
- Average bitrate depends on `quality` setting; I haven't found out default setting, but according to our experiments with opus an average output's bitrate=~32kbps;

The quality can be changed via [SFC_SET_VBR_ENCODING_QUALITY](http://www.mega-nerd.com/libsndfile/command.html#SFC_SET_VBR_ENCODING_QUALITY): `quality=1.0`(i.e. the highest quality and ~256kbps) matches `--comp 10` option. A problem is that `pysoundfile` doesn't support this command yet, so it takes some effort to extend it properly.

**Comment by snakers4** [comment]
Our experiments also confirm that using `opus` at least for inference provides the same quality as `wav`.
